### PR TITLE
Feature/create bookmarks

### PR DESCRIPTION
**Pull request by psolyca**
Some little thinks to enhanced PR https://github.com/awhipp/evetrade/pull/115 and be more accurate on issue https://github.com/awhipp/evetrade/issues/94 reply.

### CODE REVIEW

**Comment by psolyca** [comment]
@neocryptek iti s what you ask for !
@timegrinder if you need to test https://deploy-preview-127--evetrade.netlify.com/ ^^

**Comment by timegrinder** [comment]
@psolyca gave it a test, works pretty well and will be very handy in the future for those who regularly use the same settings, however there's some settings that aren't stored.

Buy From / Sell To Order type, (Applies to Station to Station and Region to Region) - suggest adding parameter
**order_type=** with parameters 's2b' (default) 's2s', 'b2s', 'b2b' or 'sell2buy' 'sell2sell' etc for readability

Security Status, (region to region) suggest parameter
**security=**  with parameters 'none' (default), 'lowhigh', 'high'

Route Preference, (region to region) suggest parameter
**route=**  with parameters 'short' (default), 'safe', 'leastsafe'

Although, those suggested parameters may not be the best as I haven't reviewed the code to see how the parameters link to the input fields apart from eyeballing the live HTML.

I even tried to break it by throwing massive numbers of stations into the station to station trading window, and it even handled that well, nicely done!

**Comment by psolyca** [comment]
@timegrinder should be ok

BTW, I think of only showing values that are different from default but if one want to add a custom value, he should know the parameter which are not visible anywhere except in the code !
What do you think ?

**Comment by awhipp** [comment]
Rather than having a "Create Bookmark" checkbox -- Would it make more sense to just do this by default?

I feel that we need a hover over text to explain what "Create Bookmark" does if we have that checkbox there. I asked a friend to go and try it and they were confused by it until I told them to look at the URL.

**Comment by timegrinder** [comment]
@awhipp I agree with the doing it by default, I forgot to mention it in my previous comment after having the same thought. The checkbox would only be viable without the hover explanation if it somehow triggered the browsers 'Bookmark Page' dialogue, which should then make it a button and not a checkbox, but also kind of add an unnecessary step. Although a 'Save This Search' could be a different feature, but that would make it cookie/user based and require additional things that could also be skipped just by letting people bookmark search URL's.

@psolyca I'm of two minds for the URL stuff, because while as a user I'd like a short URL so I could easily read what it does, it's not possible because the URL becomes complicated to read once you start shortening things too much, and tends to becomes long really easily if you start adding settings or adding stations to it. 
Overall I'd suggest leaving it as 'show all the settings, even if they're defaults'. I would however say make the options that are 'unset' (ie, max cargo, max budget) as '0' in the URL if possible, because a real setting of '0' is impossible if input into the fields, so wouldn't collide with a real user setting, while also cutting down on some of the unnecessary length of the URL.

Then we just write some documentation and put it up here on GitHub to list off all the options and what the parameters are if people want to feed in pre-built requests without having to open the page themselves, I can even do that once they're all defined and the parameters are settled on etc.

**Comment by psolyca** [comment]
I can not add a button directly under the "Search" one cause I have used variables initialised by the submitted form in order to get my parameters.

Maybe if I add a button to the bar after "Copy" "CSV" "Excel" "PDF", it could be better.
It is an other mean to save the search by bookmarking it !
BTW, this button can not open the Bookmark dialog box as some browsers do not support it (Chrome, Safari and Opera).
An alert message could popup to advertise the user he can bookmark it (address has change before alert).

@timegrinder I am also in favor of a short URL. I will see if I really need a value after the "=" cause I do not like setting it to "0" even if it is not possible.
I also could use shorten parameter names as you suggested it but it could be less informative (not sure of the english word ! :p) for users.

**Comment by timegrinder** [comment]
@psolyca Yeah I wasn't suggesting actually adding the button, it was just an explanation of the alternatives and why it should be kept as just 'Always use that kind of URL'
Perhaps some text somewhere 'You can save this search for later by bookmarking it! (Ctrl + D)' 
I'm assuming the shortcut key is the same for the common browsers.

Informative is definitely the right word. Perhaps there can be two sets of parameters that both equate to the same thing? The short version or the full version could then be accepted depending on the user preference (we would include both in the documentation so that users can learn the shorthand if they want to, or otherwise use the long versions if they wish). Either of them could be made the default, depending on how you want to set it up.

The short version would then be able to be REALLY short, like 'threshold_profit' could be 't_prof' and 'threshold_roi' would be 't_roi' etc, with the long versions fully human readable.

Where the 'create bookmark' checkbox is currently could then be converted into a 'use human readable parameter names in URL' (or 'shorten URL') with a text explaining what it means for those who wish to save a human readable version.

**Comment by psolyca** [comment]
Ok for your explanation but it should be kept simple.
So :

- The checkbox could be kept but with a hover over text explaining the checkbox (hoping people read it). This checkbox create a bookmark URL in 'human readable' parameters with minimal parameters. A message (disapearring when Ctrl-D is used if possible) on the search page could inform the user like this : 

You can save this search for later by bookmarking it! (Ctrl + D)
[Github for more parameters](https://github.com/awhipp/evetrade)

- A button (just to keep my idea ^^) in the search page to save the search if omitting the checkbox. This will create a bookmark URL with shorten parameters and the same message as above !
- On Github, the list of parameters (both long and short) with explanation
- Maybe also the explanation in help page for those who do not go on Github ?

Are you both ok with this ? Please make a list to keep it concise ^^; if you want to change/add something.

**Comment by awhipp** [comment]
I'm a fan of keeping it simple. My thoughts:

1. Not sure if two sets of parameters (long and short) make sense. We should just stick with one convention and make sure it's either human readable or documented if its not. More risk for bugs by translating extra parameters.
2. I think that the bookmark feature is a great one and should just be done by default. I do not think a checkbox is needed for bookmarks. Users should be able to just create a bookmark in their browser (`CTRL + D` or otherwise).

I'll defer to the both of you to decide though.

**Comment by timegrinder** [comment]
I can understand the simplicity side of things, I can definitely get behind a single set of parameters, would suggest it be the human readable one (but still be documented). If the URL's going to be 'long ish' anyway, making it longer won't be a problem. We're not likely to run into the ~2000 character limit for URL's unless someone tries to add an obscene number of stations to a 'Station to Station' search, as the basic 'Region to Region' ends up <300. If they do hit the character limit, they're doing something wrong to begin with.

I also agree that the parameters should just always be used in the URL instead of being via a checkbox to add them, and have a 'You can save this search by bookmarking it!' notification somewhere obvious on the page. Then people can copy the link, or bookmark it, without having to do anything extra. Means that there's 'one way' to handle the URL. There's no issue with it being too long if people don't want to save the search, because they're not likely to care about the URL if it's one off searches anyway.

There's really no need to give an API-like interface to requesting searches but then make it secondary/optional when I would think we really actually want to make it widely available and used, so having the URL parameters in by default would show many people the functionality when they might otherwise miss it, also simplifying the UI by not adding extra buttons.

**Comment by psolyca** [comment]
Ok let's go.
Less work for me ^^

**Comment by psolyca** [comment]
@timegrinder and @awhipp 
guys let's have fun ^^
Not less work as intended.
https://deploy-preview-127--evetrade.netlify.com/

I just need to write a little help for parameters on Github main page.

**Comment by awhipp** [comment]
Design looks a little broken to me. Sticks out a lot w/ the black background.

This is what I see: 
before

Thoughts on something a little more subtle like this?
after

I just moved it under the datatable and centered it.

**Comment by timegrinder** [comment]
I would suggest keeping the 'You can save this by bookmarking it' (move it to down to where @awhipp 
 suggested, and change the colours though, having it stand out is good, but maybe not stand out THAT much. Suggest giving it 33-50% opacity black background) and possibly move the parameters information link to the 'Help' section in the top right.

Can we keep any common parameters named the same for ease of use? Many of the parameters are ones that could be the same no matter what search type was being used. This saves people from having to change the simple parameter names if they want to change the trade type (ie having to change from 'region-profit-threshold' to 'profit-threshold' when switching between Region and Station trading.

For example:
(old) 'trade=r2r&threshold_roi=2&threshold_profit=2&sales_tax=2.8'
vs
(current) 'trade=2&region-roi-threshold=2&region-profit-threshold=2&region_sales_tax=2.8'

If we can keep the 'trade' parameter human readable (s2s, r2r, sst), and the common parameters kept the same (threshold_roi, threshold_profit, sales_tax etc). Also keep all of the parameter 'spaces' to underscores instead of switching between _ and - .

A suggested list of parameter names (and options) that would be common between searches, or not needed in some searches, as well as making them more human readable:

**parameter (option, option, option)**
trade_type (sst, s2s, r2r)
start (string list of stations)
end (string list of stations)
buy_type (sell / buy)
sell_type (sell / buy)
sales_tax (float)
min_roi (int) (min for minimum, more people would understand that than threshold)
min_profit (int) (as above)
max_cargo (float or int, the game does float though)
max_budget (int)
min_security (none, low, high)
route_type (none, safe, least_safe)
include_citadels (y)

Once we have the parameters nailed down I can write the bulk of the parameters documentation and save you some work :)

**Comment by psolyca** [comment]
@awhipp I had just make a warning box which fade away after 20s and can be definitly hidden. Just to warn people who know the website there is a new functionality.
BTW, I can make it as you wish ^^

@timegrinder I know about parameters names not homogeneous, it is @awhipp fault :p
It tried to simplify as much as possible the code. Parameters names are the ID of each input in the form and as each input must have a different ID, there is lots of variables.

So,

-  for my sanity, I will homogenize IDs and variables names to underscore 'space' separator ^^ (another PR !)
- link IDs to a simple variable name as you suggested it <suggestion_box>and rename internal variables to these also for homogenization - I like this word ^^ - and futur update if @awhipp is ok ^^; - I do not want to upset the master (another PR !)</suggestion_box>
- revert trade parameter to "s2s", "r2r" and "sst" (was thinking of it also)
- move bookmark warning at bottom.

**Comment by psolyca** [comment]
@awhipp and @timegrinder I do not know if you had time to look at the PR with the new commits ?

**Comment by timegrinder** [comment]
Hey, sorry, I'd forgotten to give it a test. Looks really good so far.

What is 'sales_tax_other' parameter? Is it needed? Oh, is it for when someone puts in 'Other' and manually enters a number? Could we not just have that set the 'sales_tax' parameter and leave out the '_other' altogether for cleanliness?

**Comment by psolyca** [comment]
Yes it is.
Maybe I can check against the predefined values ! If it is not one of this value, it is 'Other'.

**Comment by timegrinder** [comment]
I was more thinking the predefined value should just be a menu option that sets a float number (ie, the sales_tax parameter), with 'Other' allowing manual user input of the number.

If the user chooses 2.8 from the list, the number is set as 2.8, if the user chooses 'Other' and puts in a number, it's whatever that number is.

At least that's how I assumed it worked, can't say I've read the code.

**Comment by psolyca** [comment]
I was thinking of the URL parameters to avoid the "_other" argument.
I will not change the interface ^^

**Comment by timegrinder** [comment]
That's cool, I think the end result is the same anyway if you just make the 'sales_tax' parameter get set and don't have the _other parameter.

**Comment by psolyca** [comment]
@timegrinder Should be better ^^

**Comment by awhipp** [comment]
@psolyca Ill try to review this today or over the weekend. Unfortunately I typically work 50-60 hour weeks so I cannot get to projects like this until the weekends. thanks for all the work. If @timegrinder has not found issues I do not anticipate I will nitpick much.

**Comment by psolyca** [comment]
@awhipp take your time. I know what is working so much even if I don't anymore.

**Comment by awhipp** [comment]
Looks good from prelim tests. Merged. I'll test more in dev as time permits during the holidays. Thanksl
